### PR TITLE
Make Single Select Act More Natively

### PR DESCRIPTION
**Pull request by gojohnnygo**
## Overview
This PR makes single select act more natively.

## Issue
https://github.com/JedWatson/react-select/issues/948#issuecomment-289309520

## Considerations
1. Because core functionality has changed, there are some breaking tests. I didn't want to take time to try to fix them if this new functionality was rejected.
2. There was a CSS fix to correct a pixel jump when switching from input to value label and vice versa. I tested on the latest Mac Safari, Chrome, and FF. How far back do we need to test? What about IE?

#### New
---
new-focus
#### Old
---
old-focus

#### New
---
new-click-inside

#### Old
---
old-click-inside

#### New
---
new-click-outside

#### Old
---
old-click-outside

### CODE REVIEW

**Comment by gojohnnygo** [comment]
Any feedback on this? :)

Like I said, there are breaking tests bc things have are fundamentally changed. I don't want to try to fix them if people don't agree with the changes.

File: src/Select.js [hunk 1]
```diff
@@ -339,31 +340,38 @@ const Select = React.createClass({
 			});
 		}
 
+		const inputLabel = this.getInputLabel();
+
+		// determines if the user mouse downed on the actual ValueComponent excluding placeholder text
+		const isOnValue = this.valueComponent && (this.valueComponent.valueNode === event.target) ? true : false;
```
Comment (agirton) [inline src/Select.js]: This probably could be simplified to just be `this.valueComponent && (this.valueComponent.valueNode === event.target)`

File: src/Select.js [hunk 2]
```diff
@@ -339,31 +340,38 @@ const Select = React.createClass({
...
+
+		// clears value so that the cursor will be a the end of input then the component re-renders
+		this.input.getInput().value = '';
```
Comment (agirton) [inline src/Select.js]: Just like the check below we should probably check `getInput` is a function before we access it.

File: src/Select.js [hunk 3]
```diff
@@ -339,31 +340,38 @@ const Select = React.createClass({
...
 				isOpen: true,
 				isPseudoFocused: false,
+				inputValue: inputLabel
+			}, () => {
+				isOnValue && this.input.select();
 			});
 		} else {
 			// otherwise, focus the input and open the menu
-			this._openAfterFocus = this.props.openOnFocus;
+			this._openAfterFocus = true;
```
Comment (agirton) [inline src/Select.js]: Why was this removed?

File: src/Select.js [hunk 4]
```diff
@@ -165,6 +165,7 @@ const Select = React.createClass({
 
 	getInitialState () {
 		return {
+			hasInputChange: false,
```
Comment (agirton) [inline src/Select.js]: Nit: to follow other naming schemes `hasInputChanged` would probably be better.

File: src/Select.js [hunk 5]
```diff
@@ -415,13 +425,20 @@ const Select = React.createClass({
...
-		var isOpen = this.state.isOpen || this._openAfterFocus || this.props.openOnFocus;
+		const isOpen = this.state.isOpen || this._openAfterFocus || this.props.openOnFocus;
 		if (this.props.onFocus) {
 			this.props.onFocus(event);
 		}
+
+		// To make the react-select act more natively, render the input value with the corresponding label on focus. However, we 
+		// cannot do this for custom valueRenderer's bc the label can be manipulated during render (e.g. option.label.toUpperCase())
+		// so the option labels could be inconsistent.
+		let inputLabel = this.getInputLabel();
```
Comment (agirton) [inline src/Select.js]: It doesn't look like this is reassigned or used elsewhere. Could you just call this when update the state on line 441?

Comment (agirton) [inline src/Select.js]: Also it seems like users with valueRenderer should be able to benefit from this. Maybe we can warn users against manipulating the value?

File: scss/control.scss [hunk 6]
```diff
@@ -147,14 +147,22 @@
...
+		padding: -1 (($select-input-internal-height - 14) / 2 - 1) 0 (($select-input-internal-height - 14) / 2);  /* For IE 8 compatibility */
 		-webkit-appearance: none;
 
 		.is-focused & {
 			cursor: text;
 		}
 	}
+}
 
+// in FF, the input field jumps a pixel when enabled, this adjust for that so the label and input are perfectly aligned
```
Comment (agirton) [inline scss/control.scss]: Using vendor specific hacks should not be used. If this selector ever goes away then the original issue may reappear. To ensure they are perfectly aligned I think we should spend a bit more time digging into why they aren't instead of adding a band aid :smile:

Comment (gojohnnygo) [reply scss/control.scss]: @agirton I don't disagree with you, but: https://giphy.com/gifs/13XW2MJE0XCoM0

I tried to find a fix that's not vendor specific before your comment but didn't get anywhere. I would welcome any assist from the community. :D

File: src/Select.js (hunk 3 above)
Comment (gojohnnygo) [reply src/Select.js]: Good point.

File: src/Select.js (hunk 5 above)
Comment (gojohnnygo) [reply src/Select.js]: @agirton Yeah, I think a warn makes sense. I removed `valueRender` from the `if`, but depending on what the users' func returns, it could blow up.

**Comment by JedWatson** [comment]
Thanks for this @gojohnnygo 

I love the effort you've gone to with the PR, but am not going to merge as-is because it's a major behaviour change for react-select. The ideas are good, so my next goal with this project is to work out how to make react-select more flexible (and have less props) and it would be great to revisit how this could be made an option once that's in place.

**Comment by bluespore** [comment]
@JedWatson Any update on whether the above can be achieved through configuration now? These changes are exactly what I'm looking for and I'm keen to understand how to integrate.
